### PR TITLE
SampleDataEvent

### PR DESCRIPTION
**Pull request by X-sam**
This is a freshened up version of #1723. 

Quoting the original PR:
> Implementation of Flash's SampleDataEvent using WebAudio on HTML5 and OpenAL on Windows/Neko.
> Though untested it should work on other targets using on OpenAL.

In the time since the original PR was proposed, Chrome has [banned autoplaying audio](https://developer.chrome.com/blog/autoplay/#webaudio), so here's a modified version of that test program supporting modern browser & desktop via neko/hl targets:
```haxe
package;

import openfl.display.BitmapData;
import openfl.events.MouseEvent;
import openfl.events.Event;
import openfl.display.InteractiveObject;
import openfl.display.SimpleButton;
import openfl.display.Bitmap;
import openfl.Assets;
import openfl.display.Sprite;
import openfl.media.Sound;
import openfl.events.SampleDataEvent;
import openfl.utils.ByteArray;

class Main extends Sprite {
	private var sampleRate:Int;
	private var bufferSize:Int = 8192;
	private var bpm:Int = 90;
	private var numberOfRows:Int = 48; // 48=3/4 time | 64=4/4 time
	private var currentRow:Int = 0;
	private var quarterNoteLength:Float;
	private var sixteenthNoteLength:Float;
	private var numOctaves:Int = 8;
	private var patterns:Array<Dynamic> = new Array();
	private var currentPattern:Int = 0;
	private var songOrder:Array<Int> = [0, 1];
	private var notes:Array<String> = ["c-", "c#", "d-", "d#", "e-", "f-", "f#", "g-", "g#", "a-", "a#", "b-"];
	private var frequencies:Array<Dynamic> = new Array();
	private var samplePosition:Int = 0;
	private var position:Int = 0;
	private var channel1:Dynamic = {
		volume: .05,
		waveForm: "",
		frequency: [],
		noteTriggered: false,
		envelopePos: 0
	};
	private var envelope:Array<Float> = new Array();
	private var echoBytes:ByteArray = new ByteArray();
	private var maxEchoBytes:UInt = 8192 * 256;
	private var echoing:Bool = true;
	private var echoDelay:Float;
	private var echoPosition:UInt = 0;

	public function new() {
		super();

		var bitmapData = new BitmapData(120, 80, false, 0x010203FF);
		var pressButton = new BitmapData(120, 80, false, 0x008800FF);
		var bitmap = new Bitmap(bitmapData);
		var pressbitmap = new Bitmap(pressButton);
		var button = new SimpleButton(bitmap, bitmap, pressbitmap, bitmap);
		button.x = (stage.stageWidth - bitmap.width) / 2;
		button.y = (stage.stageHeight - bitmap.height) / 2;
		addChild(button);
		button.addEventListener(MouseEvent.CLICK, startSound);
	}

	function startSound(e:MouseEvent) {
		var sound:Sound = new Sound();
		sampleRate = sound.sampleRate;

		quarterNoteLength = sampleRate * 60 / bpm;
		sixteenthNoteLength = quarterNoteLength / 2 / 2;
		echoDelay = sixteenthNoteLength * 4;
		for (a in 0...numOctaves) {
			for (b in 0...notes.length) {
				frequencies.push([notes[b % notes.length] + a, 16.35 * Math.pow(2, frequencies.length / 12)]);
			}
		}
		patterns.push([
			"g-2|g-4", "", "", "a-4", "a#4", "", "d-2|a-4", "", "g-4", "", "f-4", "", "g-2|g-4", "", "d-4", "", "", "", "", "", "", "", "d-4", "",
			"g-2|g-4", "", "", "a-4", "a#4", "", "c-2|c-5", "", "a#4", "", "c-5", "", "d-2|d-5", "",    "", "", "", "", "", "", "", "", "d-5", ""
		]);
		patterns.push([
			"g-2|d-5", "", "", "d#5", "d-5", "", "f-2|c-5", "", "a#4", "", "a-4", "", "d#2|a#4", "", "c-5", "", "a#4", "", "d-2|a-4", "", "", "", "d-4", "",
			"g-2|g-4", "", "", "a-4", "g-4", "", "d-2|f-4", "", "d-4", "", "f-4", "", "g-2|g-4", "",    "", "",    "", "",        "", "", "", "",    "", ""
		]);
		Reflect.setProperty(channel1, "waveForm", "sawtooth");

		for (c in 0...Std.int(quarterNoteLength)) {
			if (c < sixteenthNoteLength * 1) {
				envelope.push(1.0);
			} else {
				if (c < sixteenthNoteLength * 3) {
					envelope.push(.4);
				} else {
					envelope.push(0.0);
				}
			}
		}

		updateRow();
		sound.addEventListener(SampleDataEvent.SAMPLE_DATA, onSampleData);

		sound.play();
	}

	private function updateRow():Void {
		var tempNote:String = patterns[songOrder[currentPattern]][currentRow];
		if (tempNote != "") {
			var tempArray:Array<Float> = new Array();
			Reflect.setProperty(channel1, "frequency", []);
			if (tempNote.indexOf("|") == -1) {
				// single note
				tempArray.push(findFrequency(tempNote));
			} else {
				// chord
				var tempNotes:Array<String> = tempNote.split("|");
				for (a in 0...tempNotes.length) {
					tempArray.push(findFrequency(tempNotes[a]));
				}
			}
			Reflect.setProperty(channel1, "noteTriggered", true);
			Reflect.setProperty(channel1, "frequency", tempArray);
		}
	}

	private function onSampleData(event:SampleDataEvent):Void {
		var tempArray:Array<Float>;
		var addDataL:Float = 0.0;
		var addDataR:Float = 0.0;
		var volumeModifier:Float = 0.0;
		trace("Here?");
		var sampleData:Float = 0.0;
		for (i in 0...bufferSize) {
			if (++samplePosition == sixteenthNoteLength) {
				if (++currentRow == numberOfRows) {
					currentRow = 0;
					if (++currentPattern == songOrder.length) {
						currentPattern = 0;
					}
				}
				updateRow();
				samplePosition = 0;
			}
			if (Reflect.field(channel1, "noteTriggered")) {
				Reflect.setProperty(channel1, "envelopePos", 0);
				Reflect.setProperty(channel1, "noteTriggered", false);
			}
			if (Reflect.field(channel1, "envelopePos") + 1 < envelope.length) {
				var tempInt:Int = Reflect.field(channel1, "envelopePos") + 1;
				Reflect.setProperty(channel1, "envelopePos", tempInt);
			}
			volumeModifier = envelope[Reflect.field(channel1, "envelopePos")];
			tempArray = Reflect.field(channel1, "frequency");
			if (tempArray.length == 1) {
				sampleData = generate(Reflect.field(channel1, "waveForm"), position, tempArray[0], Reflect.field(channel1, "volume") * volumeModifier);
			} else {
				sampleData = generate(Reflect.field(channel1, "waveForm"), position, tempArray[0], Reflect.field(channel1, "volume") * volumeModifier);
				sampleData += generate(Reflect.field(channel1, "waveForm"), position, tempArray[1], Reflect.field(channel1, "volume") * volumeModifier);
			}

			if (echoing) {
				if (position > echoDelay) {
					var oldPos:UInt = echoBytes.position;
					echoBytes.position = echoPosition;
					addDataL = echoBytes.readFloat() / 4;
					addDataR = echoBytes.readFloat() / 4;

					if (echoPosition + 8 < maxEchoBytes) {
						echoPosition += 8;
					} else {
						echoPosition = 0;
					}
					echoBytes.position = oldPos;
				}
			}

			event.data.writeFloat(sampleData + addDataL);
			event.data.writeFloat(sampleData + addDataR);
			echoBytes.writeFloat(sampleData);
			echoBytes.writeFloat(sampleData);
			if (echoBytes.position == maxEchoBytes) {
				echoBytes.position = 0;
			}
			position++;
		}
	}

	private function findFrequency(inpNote:String):Float {
		var retVal:Float = 0.0;
		for (a in 0...frequencies.length) {
			if (frequencies[a][0] == inpNote) {
				retVal = frequencies[a][1];
				break;
			}
		}
		return retVal;
	}

	private function generate(waveForm:String, pos:Int, frequency:Float, volume:Float):Float {
		var retVal:Float = 0.0;
		switch (waveForm) {
			case "square":
				retVal = Math.sin((pos) * Math.PI * 2 / sampleRate * frequency) > 0 ? volume : -volume;
			case "sine":
				retVal = Math.sin((pos) * Math.PI * 2 / sampleRate * frequency * 2) * volume;
			case "sawtooth":
				retVal = (2 * (pos % (sampleRate / frequency)) / (sampleRate / frequency) - 1) * volume;
		}
		return retVal;
	}
}
```

### CODE REVIEW

**Comment by EliteMasterEric** [comment]
I've been using this for a recent project and it works exactly as expected. I'm hoping this can get merged eventually as it's a great feature.

**Comment by TerryCavanagh** [comment]
I'll second that! At some point I'd love to port my flash based audio tool Bosca Ceoil to haxe, and I've been unable to because SampleDataEvent doesn't work outside of Flash targets. I have my eyes on this 👀

**Comment by TerryCavanagh** [comment]
I have this relevant thread bookmarked, which seems to collide here, and which was my most promising lead on this: https://community.openfl.org/t/is-sampledataevent-implemented-for-windows/9222/9

**Comment by X-sam** [comment]
> I'll second that! At some point I'd love to port my flash based audio tool Bosca Ceoil to haxe, and I've been unable to because SampleDataEvent doesn't work outside of Flash targets. I have my eyes on this 👀

It was the exact process of [porting Bosca Ceoil to haxe](https://github.com/X-sam/boscaceoil) that led me to needing this, actually!  
I'm convinced there's still some underlying implementation quirks with the approach in this PR but, well, _some_ functionality is better than none!

**Comment by joshtynjala** [comment]
I have a few PRs that I plan to review after 9.2 is released. I'll add this to the list.

**Comment by TerryCavanagh** [comment]
> > I'll second that! At some point I'd love to port my flash based audio tool Bosca Ceoil to haxe, and I've been unable to because SampleDataEvent doesn't work outside of Flash targets. I have my eyes on this 👀
> 
> It was the exact process of [porting Bosca Ceoil to haxe](https://github.com/X-sam/boscaceoil) that led me to needing this, actually! I'm convinced there's still some underlying implementation quirks with the approach in this PR but, well, _some_ functionality is better than none!

Oh *whoa*!

We should talk - I'd love to merge your fork as a starting point of a future Bosca Ceoil v3.0 update. Please get in touch! (email address removed)

**Comment by joshtynjala** [comment]
I'd like to get this merged soon. Probably into the 9.3.0-Dev branch.

I gave your test program a try today. It works great on C++/Neko/HL, and in Flash/AIR too. However, when I tried the html5 target, I always got the following exception (same exception in Safari, Firefox, and Chrome):

> To be consistent with flash the listener function registered to SampleDataEvent has to provide 2048, 4096 or 8192 samples if targeting HTML5.

I checked the value of `bufferSize` at the time it is checked, and it appears to be 16384.

Haxe 4.2.5
Lime 8.0.2
macOS 13.4.1 (Apple Silicon)

**Comment by joshtynjala** [comment]
I suspect that the removal of `/ 2` in the calculation of `bufferSize` from commit fc0ec1acd3a95e99c25fa07f2fcb6ed0795c67df is the reason for the exception. I added it back in, and the sound plays for me. However, I assume that there are probably more changes required if that division is restored, and I don't know much about sound generation, so I'll leave it up to you to provide the proper fix. Hopefully, this little clue helps.

**Comment by scanline** [comment]
Let me try to shed some light into this as I'm the one who come up with the original pull request six years ago.
The `SampleDataEvent` fires, as soon as the underlying Sound object runs out of samples. Well what does that mean?

The Flashplayer itself has a hardcoded sampleRate of **44100** hz. So for one second of audio you need 44100 individual
samples - or floats so to speak. So if you write 2048 samples to the event's data property, the event will fire again
in roughly 46 milliseconds (2048/44100). This is it's **bufferSize**. As in the glory days of flash not every computer was
capable keeping up with that rate you could provide more samples at once. In fact, it can be any integer between 2048 and **8192**.
With the highest number the event fires again in ~185ms. This gives weak hardware time to breathe and lowers the risk
of audio glitches but also introduces a phenomenom called latency as hearable changes to the sound can only occur
every 185ms.

As this commit targets HTML5 I had to make some decisions to make it compatible with the `WebAudio` API.
First and foremost it's equivalent `createScriptProcessor` method also has the concept of a buffer. Unfortunately
you can't have an arbitrary number of samples there. It has to be any of the following numbers:
256, 512, 1024, 2048, 4096, 8192 or 16384

So to be compatible with flash, the only options are 2048, 4096 and 8192, which is the reason for:

> To be consistent with flash the listener function registered to SampleDataEvent has to provide 2048, 4096 or 8192 samples if targeting HTML5.

Unlike flash, you also can't simply change the buffer size at runtime. This would need to create a new AudioContext each time and is simply not provided by the underlying OpenFl/Lime.

The second problem is the aforementioned samplerate. Nowadays the device's audio hardware and the browser itself offer
higher sample rates. The de facto standard is at least 48000hz. This led to the addition of the property `sampleRate` to `openfl.media.Sound`, which is not present in the standard `flash.media.sound` API. For flash it returns a hardcoded 44100 and for HTML5 whatever rate the browser picked e.g. 48000.

So to summarize: yes, the removal of **/ 2** in
`var bufferSize:Int = Std.int(data.length / 4 / 2);`
is simply nonsense.

**Comment by joshtynjala** [comment]
I have merged this PR into the 9.3.0-Dev branch. Thank you both @scanline and @X-sam.

However, I skipped the following commit based on my understanding of the most recent reply from @scanline: https://github.com/openfl/openfl/pull/2515/commits/fc0ec1acd3a95e99c25fa07f2fcb6ed0795c67df

I confirmed again that the provided sample code is working correctly on various targets, including html5/cpp/neko/hl/air/flash. Sorry to everyone for the long delay in merging this and the original #1723. It will be included in 9.3. I'm hoping to do that release soon-ish, but it's certainly possible that 9.2.x will continue with another bugfix release first.

**Comment by TerryCavanagh** [comment]
Thanks, @joshtynjala!

I'll make some time later this month to try this out and report back on any issues. (This PR is the key to porting SiON to Haxe, which I'm really keen to do!)
